### DISCUSSION TITLE
There was a problem sending the email verification

### DISCUSSION THREAD

**Discussion by zbig-t** [envkey/envkey#8] [Q&A] [answered]
Hi,

I've been trying to configure self-hosted instance since several hours now and cannot get past the e-mail verification screen. After entering my e-mail address and clicking "Next", I'm getting "There was a problem sending the email verification" exception and can see a "clientError - CERT_HAS_EXPIRED" in the "Developer tools", no matter what. Im pretty sure I have my API instance configured properly: "SMTP settings were verified successfully" and all. In fact, the Electron frontend doesn't even hit the API endpoint before throwing this error and it behaves the same way even when I shut the API down completely.

I'm on Intel macOS but have tried in the Windows 11 VM also and got identical behaviour.

Regards,
Zbig

**Comment by danenania** [comment, marked as answer]
Hi @zbig-t, this sounds like a certificate issue at the reverse-proxy/load balancer level. If you curl https://your-subdomain.your-domain.com what do you get?

**Reply by zbig-t** [reply]
Hi @danenania , thanks for your response. I don't think that is the case as I've regenerated my Let's Encrypt cert just prior to trying the app out, with a new hostname added. Please see the curl responses:

```
curl https://ks.[redacted].io
API OK%
```

```
curl -H "Upgrade: websocket" https://ks.[redacted].io
Socket Server OK%
```

Is it possible that EnvKey gets caught off-guard by my domain that is under ".io" TLD?

**Reply by danenania** [reply]
".io" shouldn't be a problem. I'm wondering if there might be an issue with the Let's Encrypt root certificate.

Just to check, does it work if you do the following:

- Close the UI.
- Then in a new terminal window:

```
envkey core stop
NODE_TLS_REJECT_UNAUTHORIZED=0 envkey core start --inline
```

- Then open the UI again and try again to verify your email.

If it works this way, it will narrow the problem down to the certificate validation.

**Reply by zbig-t** [reply]
That didn't seem to change that, unfortunately. I got this:

```
zbig@zbig-macmini ~ % envkey core stop
Core process stopped.

zbig@zbig-macmini ~ % NODE_TLS_REJECT_UNAUTHORIZED=0 envkey core start --inline
running cli_core_proc
Started EnvKey core process.
{"ts":"2022-04-28T20:17:15.051Z","msg":"Starting EnvKey core process..."}
will write to log file at  /Users/zbig/.envkey/logs/core-2022-04-28.log
{"ts":"2022-04-28T20:17:15.052Z","msg":"logger initialized with name core"}
{"ts":"2022-04-28T20:17:15.053Z","msg":"Fetching root-device-key from OS credential store..."}
{"ts":"2022-04-28T20:17:15.053Z","msg":"CLI inside `pkg`"}
{"ts":"2022-04-28T20:17:15.053Z","msg":"loading keytar from ./envkey-keytar.node"}
{"ts":"2022-04-28T20:17:15.066Z","msg":"keytar was loaded"}
{"ts":"2022-04-28T20:17:15.183Z","msg":"Fetched root-device-key."}
{"ts":"2022-04-28T20:17:15.186Z","msg":"Fetching local-state from file store..."}
{"ts":"2022-04-28T20:17:15.366Z","msg":"Fetched and decrypted local-state."}
{"ts":"2022-04-28T20:17:15.370Z","msg":"Refreshing signed in sessions.","firstToFetch":[],"restToFetch":[]}
{"ts":"2022-04-28T20:17:15.371Z","msg":"Starting local socket server on port 19048..."}
{"ts":"2022-04-28T20:17:15.374Z","msg":"EnvKey local server is running. Server port: 19047. Websocket port: 19048."}
{"ts":"2022-04-28T20:17:15.374Z","msg":"Started core process","port":"19047","wsport":"19048"}
{"ts":"2022-04-28T20:17:24.353Z","msg":"GET /state"}
{"ts":"2022-04-28T20:17:24.379Z","msg":"Websocket connection successful.","ip":"::1"}
{"ts":"2022-04-28T20:17:24.400Z","msg":"GET /state"}
{"ts":"2022-04-28T20:17:33.759Z","msg":"GET /stop"}
{"ts":"2022-04-28T20:17:33.760Z","msg":"Exiting due to SIGTERM."}
{"ts":"2022-04-28T20:17:33.760Z","msg":"Shutting down gracefully..."}
{"ts":"2022-04-28T20:17:33.760Z","msg":"Persisting final state before shutdown..."}
{"ts":"2022-04-28T20:17:33.761Z","msg":"Comitting local-state to file store..."}
{"ts":"2022-04-28T20:17:33.785Z","msg":"local-state encrypted and committed."}
{"ts":"2022-04-28T20:17:33.785Z","msg":"Finished persisting state. Shutting down."}
zbig@zbig-macmini ~ % curl localhost:19047
{"error":"Unauthorized"}%
```

Then tried the UI again but got the CERT_HAS_EXPIRED there again. Please note how it says "Shutting down" few seconds after it started, before I got any chance to do anything in the UI. But then again, something still responds on port 19047. Is it supposed to work like that?

**Reply by danenania** [reply]
Seems that the UI is stopping it and restarting its own core process, which it shouldn't be doing if the versions line up. We'll take a look at that. The 'unauthorized' error from curling localhost:19047 is expected.

I think I may see the cause of the original issue. It seems CLI/core process (which makes the request) could be on a version of node 16 that doesn't include the most recent LetsEncrypt CA. Assuming that's the issue, we should be able to get a fix out soon. I'll keep you posted.

**Reply by zbig-t** [reply]
Cool, thank you.

I've tried a different method of passing environment variables to macOS apps: `launchctl setenv NODE_TLS_REJECT_UNAUTHORIZED 0`. That seems to make the var available globally, to all new processes. This time, after I relaunched EnvKey UI, I got the signup email.

For what it's worth, I have also tried this: `launchctl setenv NODE_EXTRA_CA_CERTS "/Users/zbig/Downloads/cacert.pem"` (while keeping TLS verification enabled) where cacert.pem is the latest CA cert bundle I got from https://curl.se/ca/cacert.pem but this didn't do it. I reckon if there's already an expired LE CA cert in Node's internal bundle, it takes precedence over any updated ones or something.

Anyway, looking forward to your update. And now, to explore EnvKey 😉

**Reply by zbig-t** [reply]
Quick update: commandline `envkey` tool and GUI keep fighting and terminating each other so I guess I'll still have to wait until this gets sorted out, too.

**Reply by zbig-t** [reply]
One last thing: the Dockerfile bundled in api.zommunity.zip doesn't include NODE_ENV variable yet the app will refuse to run without one being set (and the wiki page doesn't mention it in the self-hosted how-to).

**Reply by danenania** [reply]
Ok, new versions of the CLI and UI have been released that fix the core process issue and hopefully also the certificate issue. Let me know if that does the trick.

And thanks, will also add the missing NODE_ENV.

**Reply by zbig-t** [reply]
Wow, that was fast 😄 I can confirm that worked a treat. Both issues are now solved for me: no more core process restarts and the UI now properly talks to the API over TLS.

Thank you very much!

**Reply by danenania** [reply]
Good to hear @zbig-t!

**Reply by danenania** [reply]
Also just released Community API 2.0.19 which now has `NODE_ENV=production` baked in.

**Reply by zbig-t** [reply]
Nice!
